nixops/resources/aws_vpn_gateway.py

```diff
@@ -0,0 +1,155 @@
+    """Definition of an AWS VPN gateway."""
+
+    @classmethod
+    def get_type(cls):
+        return "aws-vpn-gateway"
+
+    @classmethod
+    def get_resource_type(cls):
+        return "awsVPNGateways"
+
+    def show_type(self):
+        return "{0}".format(self.get_type())
+
+class AWSVPNGatewayState(nixops.resources.ResourceState, EC2CommonState):
+    """State of a AWS VPN gateway."""
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    access_key_id = nixops.util.attr_property("accessKeyId", None)
+    _reserved_keys = EC2CommonState.COMMON_EC2_RESERVED + ["vpnGatewayId"]
+
+    def __init__(self, depl, name, id):
+        nixops.resources.ResourceState.__init__(self, depl, name, id)
+        self._state = StateDict(depl, id)
+        self._client = None
+        self.handle_create_vpn_gtw = Handler(['region', 'zone', 'vpcId'], handle=self.realize_create_vpn_gtw)
+        self.handle_tag_update = Handler(['tags'], after=[self.handle_create_vpn_gtw], handle=self.realize_update_tag)
+
+    @classmethod
+    def get_type(cls):
+        return "aws-vpn-gateway"
+
+    def show_type(self):
+        s = super(AWSVPNGatewayState, self).show_type()
+        if self._state.get('region', None): s = "{0} [{1}]".format(s, self._state['region'])
+        return s
+
+    @property
+    def resource_id(self):
+        return self._state.get('vpnGatewayId', None)
+
+    def prefix_definition(self, attr):
+        return {('resources', 'awsVPNGateways'): attr}
+
+    def get_defintion_prefix(self):
+        return "resources.awsVPNGateways."
+
+    def connect(self):
+        if self._client: return
+        self._client = nixops.ec2_utils.connect_ec2_boto3(self._state['region'], self.access_key_id)
+
+    def create_after(self, resources, defn):
+        return {r for r in resources if
+                isinstance(r, nixops.resources.vpc.VPCState)}
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        diff_engine = self.setup_diff_engine(config=defn.config)
+
+        self.access_key_id = defn.config['accessKeyId'] or nixops.ec2_utils.get_access_key_id()
+        if not self.access_key_id:
+            raise Exception("please set 'accessKeyId', $EC2_ACCESS_KEY or $AWS_ACCESS_KEY_ID")
+
+        for handler in diff_engine.plan():
+            handler.handle(allow_recreate)
+
+    def realize_create_vpn_gtw(self, allow_recreate):
+        config = self.get_defn()
+
+        if self.state == self.UP:
+            if not allow_recreate:
+                raise Exception("VPN gateway {} defintion changed"
+                                " use --allow-recreate if you want to create a new one".format(
+                                    self._state['vpnGatewayId']))
+            self.warn("VPN gateway changed, recreating...")
+            self._destroy()
+            self._client = None
+
+        self._state['region'] = config['region']
+
+        self.connect()
```

Comment:
It would be nice to get rid of these `self.connect()` calls by replacing `self._client` with something like `self.get_client()` (which would cache the client object).